src/main/java/build/buf/protovalidate/ValidatorFactory.java

```diff
@@ -0,0 +1,159 @@
+      this.disableLazy = disableLazy;
+    }
+
+    /**
+     * Build the validator, warming up the cache with any provided descriptors.
+     *
+     * @return A Validator instance
+     * @throws CompilationException If any of the given descriptors' validation rules fail
+     *     processing while warming up the cache.
+     * @throws IllegalStateException If disableLazy is set to true and no descriptors are passed.
+     */
+    public Validator build() throws CompilationException, IllegalStateException {
+      if (disableLazy && this.descriptors.size() == 0) {
+        throw new IllegalStateException(
+            "a list of descriptors is required when disableLazy is true");
+      }
+
+      Config cfg = this.config;
+      if (cfg == null) {
+        cfg = Config.newBuilder().build();
+      }
+      return new ValidatorImpl(cfg, this.descriptors, this.disableLazy);
+    }
+
+    @Override
+    EagerBuilder self() {
+      return this;
+    }
+
+    boolean getDisableLazy() {
+      return this.disableLazy;
+    }
+
+    List<Descriptor> getDescriptors() {
+      return this.descriptors;
+    }
+  }
+
+  /** A builder class used for building a validator. */
+  public static class Builder extends BaseBuilder<Builder> {
+    // Default constructor
+    private Builder() {}
+
+    /**
+     * Build a new validator
+     *
+     * @return A new {@link Validator} instance.
+     */
+    public Validator build() {
+      Config cfg = this.config;
+      if (cfg == null) {
+        cfg = Config.newBuilder().build();
+      }
+      return new ValidatorImpl(cfg);
+    }
+
+    @Override
+    Builder self() {
+      return this;
+    }
+  }
+
+  /**
+   * A convenience function for creating a new validator with a default configuration.
+   *
+   * @return A Validator instance
+   */
+  public static Validator defaultInstance() {
+    return newBuilder().build();
+  }
+
+  /**
+   * Creates a new builder for a validator.
+   *
+   * @return A Validator builder
+   */
+  public static Builder newBuilder() {
+    return new Builder();
+  }
+
+  /**
+   * Creates a new builder for a validator.
+   *
+   * @param descriptors the list of descriptors to warm up the cache.
+   * @param disableLazy whether to disable lazy loading of validation rules. When validation is
+   *     performed, a message's rules will be looked up in a cache. If they are not found, by
+   *     default they will be processed and lazily-loaded into the cache. Setting this to false will
+   *     not attempt to lazily-load descriptor information not found in the cache and essentially
+   *     makes the entire cache read-only, eliminating thread contention.
+   * @return A Validator builder
+   */
+  public static EagerBuilder newBuilder(List<Descriptor> descriptors, boolean disableLazy) {
+    return new EagerBuilder(descriptors, disableLazy);
+  }
+}
+
+/** An abstract class that all validator builders should extend. */
+abstract class BaseBuilder<T extends BaseBuilder<T>> {
```

Comment:
This should be made static and lifted into the ValidatorFactory class or moved to a separate .java file.